framework/finalizer.go

```diff
@@ -24,30 +24,57 @@ type patchSpec struct {
 
 func (f *Framework) addFinalizer(obj interface{}) (stopReconciliation bool, err error) {
 	restClient := f.k8sClient.CoreV1().RESTClient()
-	patch, path, result, err := createAddFinalizerPatch(obj, f.name)
-	if err != nil {
-		return false, microerror.Mask(err)
-	}
-	if patch == nil {
-		return result, err
-	}
-	p, err := json.Marshal(patch)
+	// We get the accessor of the object which we got passed from the framework.
+	accessor, err := meta.Accessor(obj)
 	if err != nil {
 		return false, microerror.Mask(err)
 	}
+	path := accessor.GetSelfLink()
+	// Continuing with reconciliation is set as a default here.
+	stopReconciliation = false
+
 	operation := func() error {
-		res := restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
-		if res.Error() != nil {
+		// We get an up to date version of our object from k8s.
+		res := restClient.Get().AbsPath(path).Do()
+		// We parse the response from the RESTClient to runtime object.
+		obj, err = res.Get()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		patch, result, err := createAddFinalizerPatch(obj, f.name)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if patch == nil {
+			// patch is empty, nothing to do.
+			// We trust createAddFinalizerPatch with the decision to stop reconciliation.
+			stopReconciliation = result
+			return nil
+		}
+		p, err := json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		res = restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
+		if IsInvalidResourceVersionError(res.Error()) {
+			// We log a warning, this should not be the case. This warning should help
+			// identify race conditions.
+			f.logger.Log("function", "addFinalizer", "level", "warning", "message", "object has out of date ResourceVersion set")
+			return microerror.Mask(res.Error())
+		} else if res.Error() != nil {
 			return microerror.Mask(res.Error())
 		}
+		// The finalizer was added, we should stop reconciliation and wait for the
+		// next update event to come in.
+		stopReconciliation = true
 		return nil
 	}
 	err = backoff.Retry(operation, f.backOffFactory())
 	if err != nil {
 		return false, microerror.Mask(err)
 	}
 
-	return true, nil
+	return stopReconciliation, nil
```

Comment:
Just a general question again. Why do we do this boolean magic here?

Reply:
The event will be replayed as the object is mutated (by adding a finalizer). This is to prevent double reconciliation. This loop adds finalizer only.